src/Core/Label.js

```diff
@@ -0,0 +1,154 @@
+ * Label. Default is true. You should not change this, as it is used internally
+ * for optimisation.
+ * @property {Element} content - The DOM object that contains the content of the
+ * label. The style and the position are applied on this object.
+ * @property {THREE.Vector3} position - The position in the 3D world of the
+ * label.
+ * @property {Coordinates} coordinates - The coordinates of the label.
+ */
+class Label extends THREE.Object3D {
+    /**
+     * @param {Element|string} content - The content; can be a
+     * string, with or without HTML tags in it, or it can be an Element.
+     * @param {Coordinates} coordinates - The world coordinates, where to place
+     * the Label.
+     * @param {Style} style - The style to apply to the content. Once the style
+     * is applied, it cannot be changed directly. However, if it really needed,
+     * it can be accessed through `label.content.style`, but it is highly
+     * discouraged to do so.
+     */
+    constructor(content, coordinates, style = {}) {
+        if (content == undefined || coordinates == undefined) {
+            throw new Error('content and coordinates are mandatory to add a Label');
+        }
+
+        super();
+
+        let _visible = this.visible;
+        // can't do an ES6 setter/getter here
+        Object.defineProperty(this, 'visible', {
+            set(v) {
+                if (v != _visible) { // avoid changing the style
+                    _visible = v;
+                    this.content.style.display = v ? 'block' : 'none';
+                    // TODO: add smooth transition for fade in/out
+                }
+            },
+            get() {
+                return _visible;
+            },
+        });
+
+        this.isLabel = true;
+        this.coordinates = coordinates;
+
+        this.projectedPosition = { x: 0, y: 0 };
+        this.boundaries = { left: 0, right: 0, top: 0, bottom: 0 };
+
+        this.content = document.createElement('div');
+        this.content.style.userSelect = 'none';
+        this.content.style.position = 'absolute';
+        if (typeof content == 'string') {
+            this.content.textContent = content || '';
+        } else {
+            this.content.appendChild(content);
+        }
+        this.baseContent = content;
+
+        if (style.isStyle) {
+            this.anchor = style.getTextAnchorPosition();
+            style.applyToHTML(this.content);
+        }
+
+        this.zoom = {
+            min: style.zoom && style.zoom.min != undefined ? style.zoom.min : 2,
+            max: style.zoom && style.zoom.max != undefined ? style.zoom.max : 24,
+        };
+    }
+
+    /**
+     * Moves a label on the screen, using screen coordinates. It updates the
+     * boundaries as it moves it.
+     *
+     * @param {number} x - X coordinates in pixels, from left.
+     * @param {number} y - Y coordinates in pixels, from top.
+     */
+    preMove(x, y) {
```

Comment:
Could you rename to `updateProjectedPosition`?

Reply:
Done